fabric_generator/code_generation_Verilog.py

```diff
@@ -0,0 +1,236 @@
+from typing import Literal
+from fabric import Tile, Bel, ConfigBitMode, IO
+import math
+import re
+
+from code_generator import codeGenerator
+
+
+class VerilogWriter(codeGenerator):
+
+    def addComment(self, comment, onNewLine=False, end="", indentLevel=0) -> None:
+        if onNewLine:
+            self._add("")
+        if self._content:
+            self._content[-1] += f"{' ':<{indentLevel*4}}" + \
+                f"//{comment}"f"{end}"
+        else:
+            self._add(f"{' ':<{indentLevel*4}}" +
+                      f"// {comment}"f"{end}")
+
+    def addHeader(self, name, package='', maxFramesPerCol='', frameBitsPerRow='', ConfigBitMode='FlipFlopChain', indentLevel=0):
+        self._add(f"module {name}", indentLevel)
+
+    def addHeaderEnd(self, name, indentLevel=0):
+        pass
+
+    def addParameterStart(self, indentLevel=0):
+        self._add(f"#(", indentLevel)
+
+    def addParameterEnd(self, indentLevel=0):
+        self._add(")", indentLevel)
+
+    def addParameter(self, name, type, value, end=False, indentLevel=0):
+        if end:
+            self._add(f"parameter {name}={value}", indentLevel)
+        else:
+            self._add(f"parameter {name}={value},", indentLevel)
+
+    def addPortStart(self, indentLevel=0):
+        self._add(f"(", indentLevel)
+
+    def addPortEnd(self, indentLevel=0):
+        self._add(");", indentLevel)
+
+    def addPortScalar(self, name, io: IO, end=False, indentLevel=0):
+        t = io.value.lower()
+        if end:
+            self._add(f"{t} {name}", indentLevel)
+        else:
+            self._add(f"{t} {name},", indentLevel)
+
+    def addPortVector(self, name, io: IO, msbIndex, end=False, indentLevel=0):
+        t = io.value.lower()
+        if end:
+            self._add(f"{t} [{msbIndex}:0] {name}", indentLevel)
+        else:
+            self._add(f"{t} [{msbIndex}:0] {name},", indentLevel)
+
+    def addDesignDescriptionStart(self, name, indentLevel=0):
+        pass
+
+    def addDesignDescriptionEnd(self, indentLevel=0):
+        self._add("endModule", indentLevel)
```

Comment:
Verilog is case-sensitive so this needs to be `endmodule` - and also a good sign this all needs testing...